x/chat/types/poll.go

```diff
@@ -0,0 +1,100 @@
+package types
+
+import (
+	"errors"
+
+	types "github.com/cosmos/cosmos-sdk/codec/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// NewPoll creates a new poll
+func NewPoll(options []string) (Poll, error) {
+	poll := new(Poll)
+
+	if !checkOptions(options) {
+		return *poll, sdkerrors.Wrap(ErrInvalidPoll, "invalid options")
+	}
+	poll.Options = options
+	poll.Votes = make(map[string]*Vote)
+
+	return *poll, nil
+}
+
+// HasUserVoted checks if a user voted for a poll
+func (poll Poll) HasUserVoted(user *User) (bool, error) {
+	// Decode into an addressable user
+	chatUser, err := user.DecodeChatUser()
+	if err != nil {
+		return false, err
+	}
+
+	_, ok := poll.Votes[chatUser.Identifier()]
+	return ok, nil
+}
+
+// GetUserVote retrieves the vote of a user
+func (poll Poll) GetUserVote(user *User) (*Vote, error) {
+	// Decode into an addressable user
+	chatUser, err := user.DecodeChatUser()
+	if err != nil {
+		return nil, err
+	}
+
+	vote, ok := poll.Votes[chatUser.Identifier()]
+	if !ok {
+		return nil, errors.New("No vote found")
+	}
+
+	return vote, nil
+}
+
+// AppendVote appends a vote into the poll
+func (poll *Poll) AppendVote(vote *Vote) error {
```

Comment:
Some uses _*Poll_ and some _Poll_ while declaring the methods, maybe we can stick with one way for all.

Reply:
`*Poll` allows explicitly telling that the method is mutable for the object.

We could have `func (poll Poll) AppendVote(vote *Vote) (Poll, error)` instead but I think it's a bit less efficient